crates/driver/src/infra/observe/mod.rs

```diff
@@ -0,0 +1,176 @@
+//! This module implements the observability for the driver. It exposes
+//! functions which represent events that are meaningful to the system. These
+//! functions are called when the corresponding events occur. They log the event
+//! and update the metrics, if the event is worth measuring.
+
+use {
+    super::Mempool,
+    crate::{
+        boundary,
+        domain::{
+            competition::{
+                self,
+                solution::{self, settlement, Settlement},
+                Auction,
+                Reveal,
+                Solution,
+            },
+            quote::{self, Quote},
+            Liquidity,
+        },
+        util::http,
+    },
+    url::Url,
+};
+
+mod metrics;
+
+// TODO The idea is that some of the functions below will also update the
+// metrics. This should be fairly easy to do, just go down the list and think
+// about if the thing is worth measuring or not. This will be done in a
+// follow-up PR.
+
+/// Observe a received auction.
+pub fn auction(auction: &Auction) {
+    tracing::info!(?auction, "received auction");
+}
```

Comment:
Not a big fan of adding this additional ceremony for very simple logs.
I see that this might be nice in case an event needs extremely verbose logging or also want to increment metrics at the same time but for very simple logs (which most of these are) I would prefer to do them inline.
As a side note logging stuff here will also obscure where in which file/component the event was triggered from.
IMO having the code location helps a lot with quickly navigating logs especially for people less familiar with the code base:
<pre>
<code>
2023-06-07T07:16:15.112Z DEBUG caching_native_price_estimator: <b>shared::price_estimation::competition</b>: winning price estimate query=Query
</code>
</pre>

Reply:
I agree. I feel logging is meant to be low level, directly in the code, with convenient access to local variables. Having it directly in the module is important to control the log level too through `tracing`, although there might be other ways to accomplish that.

For metrics I see the point in having a dedicated observer more because metrics are more of an intentional output of the system. Whereas logging is whatever we find useful for debugging at the moment.

I'm willing to try it in this style too but while keeping in mind that we should re-evaluate it after a while.